File: mobile/lib/utils/status_utils.dart
import '../services/local_db.dart';

class StatusUtils {
  static final Set<String> _availableRoomStatuses = {
    'شاغرة',
    'شاغره',
    'متاحة',
    'متاح',
    'available',
    'vacant',
    'empty',
  }.map(_normalize).toSet();

  static final Set<String> _occupiedRoomStatuses = {
    'محجوزة',
    'محجوز',
    'مشغولة',
    'occupied',
    'محجوز temporarily',
    'نشط',
    'active',
    'مؤقت',
    'provisional',
  }.map(_normalize).toSet();

  static final Set<String> _activeBookingStatuses = {
    'محجوزة',
    'محجوز',
    'نشط',
    'active',
    'confirmed',
    'قيد الحجز',
    'in_progress',
    'مؤقت',
    'provisional',
  }.map(_normalize).toSet();

  /// القائمة الخام للحالات النشطة (قبل التطبيع) - لاستخدامها في استعلامات SQL
  static const List<String> activeBookingStatuses = [
    'محجوزة',
    'محجوز',
    'نشط',
    'active',
    'confirmed',
    'قيد الحجز',
    'in_progress',
    'مؤقت',
    'provisional',
  ];

  static final Set<String> _activeEmployeeStatuses = {
    'نشط',
    'active',
  }.map(_normalize).toSet();

  /// ✅ حالات الغرف تحت الصيانة — عربية + إنجليزية، مُطبَّعة (trim+lowercase).
  /// تُستخدم في isUnderMaintenance() — مصدر وحيد للحقيقة (DRY) يمنع التشتت
  /// بين 'صيانة' و 'maintenance' في عدة providers/getters.
  static final Set<String> _maintenanceRoomStatuses = {
    'صيانة',
    'maintenance',
    'under_maintenance',
    'under maintenance',
  }.map(_normalize).toSet();

  static const List<String> activeEmployeeStatuses = ['نشط', 'active'];

  /// حالات إنهاء الخدمة
  static final Set<String> _terminatedEmployeeStatuses = {
    'مفصول',
    'terminated',
    'استقالة',
    'resigned',
    'استغناء',
    'laid_off',
  }.map(_normalize).toSet();

  static const List<String> terminatedEmployeeStatuses = [
    'مفصول',
    'terminated',
    'استقالة',
    'resigned',
    'استغناء',
    'laid_off',
  ];

  static final Set<String> _provisionalStatuses = {
    'مؤقت',
    'provisional',
  }.map(_normalize).toSet();

  static String _normalize(String value) => value.trim().toLowerCase();

  static bool isRoomAvailable(String status) =>
      _availableRoomStatuses.contains(_normalize(status));

  /// ✅ هل الغرفة تحت الصيانة؟ يفحص 'صيانة' و 'maintenance' (مع variants) بشكل
  /// مُطبَّع (trim+lowercase). استخدم هذا بدلاً من `status == 'صيانة' || status ==
  /// 'maintenance'` المكرر في عدة أماكن — يضمن اتساقاً كاملاً ويمنع bugs الترجمة.
  static bool isUnderMaintenance(String status) =>
      _maintenanceRoomStatuses.contains(_normalize(status));

  static bool isRoomOccupied(String status) =>
      _occupiedRoomStatuses.contains(_normalize(status));

  static bool isActiveBooking(String status) =>
      _activeBookingStatuses.contains(_normalize(status));

  static bool isEmployeeActive(String status) =>
      _activeEmployeeStatuses.contains(_normalize(status));

  /// هل الموظف مفصول / مستغنى عنه / استقال؟
  static bool isEmployeeTerminated(String status) =>
      _terminatedEmployeeStatuses.contains(_normalize(status));

  /// تسمية عرض الحالة بالعربية مع دعم حالات إنهاء الخدمة
  static String employeeStatusLabel(String status) {
    if (isEmployeeActive(status)) return 'نشط';
    if (_normalize(status) == _normalize('مفصول') ||
        _normalize(status) == _normalize('terminated')) {
      return 'مفصول';
    }
    if (_normalize(status) == _normalize('استقالة') ||
        _normalize(status) == _normalize('resigned')) {
      return 'استقالة';
    }
    if (_normalize(status) == _normalize('استغناء') ||
        _normalize(status) == _normalize('laid_off')) {
      return 'استغناء';
    }
    if (_normalize(status) == _normalize('مجمد') ||
        _normalize(status) == _normalize('frozen')) {
      return 'مجمد';
    }
    return 'غير نشط';
  }

  static String canonicalEmployeeStatus(String status) {
    if (isEmployeeActive(status)) {
      return 'active';
    }
    if (_normalize(status) == _normalize('مفصول') ||
        _normalize(status) == _normalize('terminated')) {
      return 'terminated';
    }
    if (_normalize(status) == _normalize('استقالة') ||
        _normalize(status) == _normalize('resigned')) {
      return 'resigned';
    }
    if (_normalize(status) == _normalize('استغناء') ||
        _normalize(status) == _normalize('laid_off')) {
      return 'laid_off';
    }
    if (_normalize(status) == _normalize('مجمد') ||
        _normalize(status) == _normalize('frozen')) {
      return 'frozen';
    }
    return 'inactive';
  }

  /// تحويل الحالة الكانونية إلى عربية
  static String canonicalToArabic(String canonical) {
    switch (canonical) {
      case 'active':
        return 'نشط';
      case 'terminated':
        return 'مفصول';
      case 'resigned':
        return 'استقالة';
      case 'laid_off':
        return 'استغناء';
      case 'frozen':
        return 'مجمد';
      default:
        return 'غير نشط';
    }
  }

  /// لون الحالة
  static int employeeStatusColor(String status) {
    if (isEmployeeActive(status)) return 0xFF4CAF50; // أخضر
    if (isEmployeeTerminated(status)) return 0xFFF44336; // أحمر
    if (_normalize(status) == _normalize('مجمد')) return 0xFFFF9800; // برتقالي
    return 0xFF9E9E9E; // رمادي
  }

  static bool isProvisional(String status) =>
      _provisionalStatuses.contains(_normalize(status));

  static bool isBookingProvisional(Booking booking) =>
      isProvisional(booking.status);

  static String roomStatusForOccupancy(
    bool occupied, {
    String fallbackAvailable = 'شاغرة',
    String fallbackOccupied = 'محجوزة',
  }) {
    return occupied ? fallbackOccupied : fallbackAvailable;
  }

  static bool isBookingActive(Booking booking) =>
      isActiveBooking(booking.status);
}

File: mobile/android/app/src/main/kotlin/com/marina/marina/domain/util/StatusUtils.kt
package com.marina.marina.domain.util

/**
 * Business-rule source of truth for the free-text status strings stored on
 * rooms/bookings/employees (Arabic + English, mixed casing/whitespace in the
 * data). Ported 1:1 from the Flutter app's `lib/utils/status_utils.dart` so
 * both clients agree on what "occupied"/"active"/"terminated" mean.
 */
object StatusUtils {

    private fun normalize(status: String): String = status.trim().lowercase()

    private val availableRoomStatuses = setOf(
        "شاغرة", "شاغره", "متاحة", "متاح", "available", "vacant", "empty"
    ).map(::normalize).toSet()

    private val occupiedRoomStatuses = setOf(
        "محجوزة", "محجوز", "مشغولة", "occupied", "محجوز temporarily", "نشط", "active", "مؤقت", "provisional"
    ).map(::normalize).toSet()

    private val maintenanceRoomStatuses = setOf(
        "صيانة", "maintenance", "under_maintenance", "under maintenance"
    ).map(::normalize).toSet()

    private val activeBookingStatuses = setOf(
        "محجوزة", "محجوز", "نشط", "active", "confirmed", "قيد الحجز", "in_progress", "مؤقت", "provisional"
    ).map(::normalize).toSet()

    private val activeEmployeeStatuses = setOf("نشط", "active").map(::normalize).toSet()

    private val terminatedEmployeeStatuses = setOf(
        "مفصول", "terminated", "استقالة", "resigned", "استغناء", "laid_off"
    ).map(::normalize).toSet()

    fun isRoomAvailable(status: String): Boolean = availableRoomStatuses.contains(normalize(status))
    fun isRoomOccupied(status: String): Boolean = occupiedRoomStatuses.contains(normalize(status))
    fun isRoomUnderMaintenance(status: String): Boolean = maintenanceRoomStatuses.contains(normalize(status))
    fun isBookingActive(status: String): Boolean = activeBookingStatuses.contains(normalize(status))
    fun isEmployeeActive(status: String): Boolean = activeEmployeeStatuses.contains(normalize(status))
    fun isEmployeeTerminated(status: String): Boolean = terminatedEmployeeStatuses.contains(normalize(status))

    /** Dart canonicalEmployeeStatus (status_utils.dart l.147-166). */
    fun canonicalEmployeeStatus(status: String): String = when {
        isEmployeeActive(status) -> "active"
        normalize(status) == normalize("مفصول") || normalize(status) == "terminated" -> "terminated"
        normalize(status) == normalize("استقالة") || normalize(status) == "resigned" -> "resigned"
        normalize(status) == normalize("استغناء") || normalize(status) == "laid_off" -> "laid_off"
        normalize(status) == normalize("مجمد") || normalize(status) == "frozen" -> "frozen"
        else -> "inactive"
    }

    /** Dart employeeStatusLabel (status_utils.dart l.126-144) — Arabic display. */
    fun employeeStatusLabel(status: String): String = when {
        isEmployeeActive(status) -> "نشط"
        normalize(status) == normalize("مفصول") || normalize(status) == "terminated" -> "مفصول"
        normalize(status) == normalize("استقالة") || normalize(status) == "resigned" -> "استقالة"
        normalize(status) == normalize("استغناء") || normalize(status) == "laid_off" -> "استغناء"
        normalize(status) == normalize("مجمد") || normalize(status) == "frozen" -> "مجمد"
        else -> "غير نشط"
    }

    /** Dart canonicalToArabic (status_utils.dart l.168+). */
    fun canonicalToArabic(canonical: String): String = when (canonical) {
        "active" -> "نشط"
        "terminated" -> "مفصول"
        "resigned" -> "استقالة"
        "laid_off" -> "استغناء"
        "frozen" -> "مجمد"
        else -> "غير نشط"
    }
}
